velox/type/parser/tests/TypeParserTest.cpp

```diff
@@ -0,0 +1,209 @@
+  ASSERT_EQ(*parseType("map(bigint,bigint)"), *MAP(BIGINT(), BIGINT()));
+
+  ASSERT_EQ(
+      *parseType("map(bigint,array(bigint))"), *MAP(BIGINT(), ARRAY(BIGINT())));
+
+  ASSERT_EQ(
+      *parseType("map(bigint,map(bigint,map(varchar,bigint)))"),
+      *MAP(BIGINT(), MAP(BIGINT(), MAP(VARCHAR(), BIGINT()))));
+}
+
+TEST_F(TestTypeSignature, invalidType) {
+  VELOX_ASSERT_THROW(parseType("blah()"), "Failed to parse type [blah()]");
+
+  VELOX_ASSERT_THROW(parseType("array()"), "Failed to parse type [array()]");
+
+  VELOX_ASSERT_THROW(parseType("map()"), "Failed to parse type [map()]");
+
+  VELOX_ASSERT_THROW(parseType("x"), "Failed to parse type [x]");
+
+  // Ensure this is not treated as a row type.
+  VELOX_ASSERT_THROW(
+      parseType("rowxxx(a)"), "Failed to parse type [rowxxx(a)]");
+}
+
+TEST_F(TestTypeSignature, rowType) {
+  ASSERT_EQ(
+      *parseType("row(a bigint,b varchar,c real)"),
+      *ROW({"a", "b", "c"}, {BIGINT(), VARCHAR(), REAL()}));
+
+  ASSERT_EQ(
+      *parseType("row(a bigint,b array(bigint),c row(a bigint))"),
+      *ROW(
+          {"a", "b", "c"},
+          {BIGINT(), ARRAY(BIGINT()), ROW({"a"}, {BIGINT()})}));
+
+  ASSERT_EQ(
+      *parseType("row(\"12 tb\" bigint,b bigint,c bigint)"),
+      *ROW({"12 tb", "b", "c"}, {BIGINT(), BIGINT(), BIGINT()}));
+
+  ASSERT_EQ(
+      *parseType("row(a varchar(10),b row(a bigint))"),
+      *ROW({"a", "b"}, {VARCHAR(), ROW({"a"}, {BIGINT()})}));
+
+  ASSERT_EQ(
+      *parseType("array(row(col0 bigint,col1 double))"),
+      *ARRAY(ROW({"col0", "col1"}, {BIGINT(), DOUBLE()})));
+
+  ASSERT_EQ(
+      *parseType("row(col0 array(row(col0 bigint,col1 double)))"),
+      *ROW({"col0"}, {ARRAY(ROW({"col0", "col1"}, {BIGINT(), DOUBLE()}))}));
+
+  ASSERT_EQ(*parseType("row(bigint,varchar)"), *ROW({BIGINT(), VARCHAR()}));
+
+  ASSERT_EQ(
+      *parseType("row(bigint,array(bigint),row(a bigint))"),
+      *ROW({BIGINT(), ARRAY(BIGINT()), ROW({"a"}, {BIGINT()})}));
+
+  ASSERT_EQ(
+      *parseType("row(varchar(10),b row(bigint))"),
+      *ROW({"", "b"}, {VARCHAR(), ROW({BIGINT()})}));
+
+  ASSERT_EQ(
+      *parseType("array(row(col0 bigint,double))"),
+      *ARRAY(ROW({"col0", ""}, {BIGINT(), DOUBLE()})));
+
+  ASSERT_EQ(
+      *parseType("row(col0 array(row(bigint,double)))"),
+      *ROW({"col0"}, {ARRAY(ROW({BIGINT(), DOUBLE()}))}));
+
+  ASSERT_EQ(
+      *parseType("row(double double precision)"), *ROW({"double"}, {DOUBLE()}));
+
+  ASSERT_EQ(*parseType("row(double precision)"), *ROW({DOUBLE()}));
+
+  ASSERT_EQ(
+      *parseType("RoW(a bigint,b varchar)"),
+      *ROW({"a", "b"}, {BIGINT(), VARCHAR()}));
+
+  ASSERT_EQ(*parseType("row(array(Json))"), *ROW({ARRAY(JSON())}));
+
+  ASSERT_EQ(
+      *parseType("row(col0 row(array(HyperLogLog)))"),
+      *ROW({"col0"}, {ROW({ARRAY(HYPERLOGLOG())})}));
+
+  // Field type canonicalization.
+  ASSERT_EQ(*parseType("row(col iNt)"), *ROW({"col"}, {INTEGER()}));
+}
+
+TEST_F(TestTypeSignature, typesWithSpaces) {
```

Comment:
I guess what is not covered today is if you add a new custom type which has blankspaces in the name.

Reply:
I modified the implementation to remove the manual Prestosql types inference.
`JSON` and `Timestamp with time zone` are now registered. `Hyperloglog` is not registered.
These take the custom type path.

Reply:
See my comment above. Ideally we would like to add a custom test which is not explicitly defined in the flex file (so it's extensible).